src/actions/plugins/imswitcher.ts

```diff
@@ -0,0 +1,142 @@
+import { logger } from '../../util/logger';
+import { Message } from '../../util/message';
+import { Globals } from '../../globals';
+
+enum ModeCluster {
+  InsertLikeMode,
+  NormalLikeMode,
+}
+
+// InputMethodSwitcher change input method automatically when mode changed
+export class InputMethodSwitcher {
+  constructor(execute: (cmd: string) => Promise<string> = util.executeShell) {
+    this.execute = execute;
+  }
+
+  private savedIMKey = '';
+  private execute: (cmd: string) => Promise<string>;
+
+  public async switchInputMethod(prevMode: ModeName, newMode: ModeName) {
+    if (configuration.autoSwitchInputMethod.enable !== true) {
+      return;
+    }
+    if (!this.isConfigurationValid()) {
+      this.disableIMSwitch();
+      return;
+    }
+    // when you exit from insert-like mode, save origin input method and set it to default
+    const prevModeCluster = this.getModeCluster(prevMode);
+    const newModeCluster = this.getModeCluster(newMode);
+    if (prevModeCluster !== newModeCluster) {
+      if (newModeCluster === ModeCluster.InsertLikeMode) {
+        this.resumeIM();
+      } else if (newModeCluster === ModeCluster.NormalLikeMode) {
+        this.switchToDefaultIM();
+      }
+    }
+  }
+
+  // save origin input method and set input method to default
+  private async switchToDefaultIM() {
+    const obtainIMCmd = configuration.autoSwitchInputMethod.obtainIMCmd;
+    const rawObtainIMCmd = this.getRawCmd(obtainIMCmd);
+    if (existsSync(rawObtainIMCmd) || Globals.isTesting) {
+      try {
+        const insertIMKey = await this.execute(obtainIMCmd);
+        if (insertIMKey !== undefined) {
+          this.savedIMKey = insertIMKey.trim();
+        }
+      } catch (e) {
+        logger.error(`IMSwitcher: promise is rejected. err=${e}`);
+      }
+    } else {
+      this.showCmdNotFoundErrorMessage(rawObtainIMCmd, 'vim.autoSwitchInputMethod.obtainIMCmd');
+    }
+
+    const defaultIMKey = configuration.autoSwitchInputMethod.defaultIM;
+    if (defaultIMKey !== this.savedIMKey) {
+      this.switchToIM(defaultIMKey);
+    }
+  }
+
+  // resume origin inputmethod
+  private async resumeIM() {
+    if (this.savedIMKey !== configuration.autoSwitchInputMethod.defaultIM) {
+      this.switchToIM(this.savedIMKey);
+    }
+  }
+
+  private async switchToIM(imKey: string) {
+    let switchIMCmd = configuration.autoSwitchInputMethod.switchIMCmd;
+    const rawSwitchIMCmd = this.getRawCmd(switchIMCmd);
+    if (existsSync(rawSwitchIMCmd) || Globals.isTesting) {
+      if (imKey !== '' && imKey !== undefined) {
+        switchIMCmd = switchIMCmd.replace('{im}', imKey);
+        try {
+          await this.execute(switchIMCmd);
+        } catch (e) {
+          logger.error(`IMSwitcher: promise is rejected. err=${e}`);
+        }
+      }
+    } else {
+      this.showCmdNotFoundErrorMessage(rawSwitchIMCmd, 'vim.autoSwitchInputMethod.switchIMCmd');
+    }
+  }
+
+  private getModeCluster(mode: ModeName): ModeCluster {
```

Comment:
This is me nitpicking, but there will only ever be two values in this enum correct? If that is the case, I think you can get rid of the enum and instead return a boolean:

```
private isInsertLikeMode(mode: ModeName): boolean
```

That should simplify the number of lines of code:

```
const isPrevModeInsertLike = this.isInsertLikeMode(prevMode);
const isNewModeInsertLike = this.isInsertLikeMode(newMode);
if (isPrevModeInsertLike !== isNewModeInsertLike ) {
    if (isNewModeInsertLike ) {
       this.resumeIM();
     } else {
        this.switchToDefaultIM();
     }
```

Is that right? Hopefully this suggestion makes sense.